### PR TITLE
Implement one-liner coregistration

### PR DESCRIPTION
**Pull request by adehecq**
This is mainly to fix issue #257.
See [David](https://github.com/dshean/demcoreg/blob/master/demcoreg/dem_align.py) and [old own](https://github.com/GeoUtils/geoutils/blob/master/geoutils/dem_coregistration.py) implementations for ideas.

This PR adds a function in `coreg.py` to run all the coregistration steps in one line, i.e.:
- load the datasets and reproject on the same grid
- create the inlier mask by excluding glaciers, steep slopes and large blunders
- estimate the coregistration: setup so that it can be split into a separate horizontal alignment (with method set by string, default is 'nuth_kaab') and a vertical alignment (default is 'median', i.e remove a median bias).
- apply the coregistration
- optionally plot a figure
- calculate the statistics of before and after coreg
- returns the coregistered DEM (optionally save to file) and statistics

To be added:
- [ ] add tests
- [ ] allow inputs to be DEM instance rather than path to file

### CODE REVIEW

**Review by rhugonnet — APPROVED** [review]
Nice to have a first version of this done! :smile:

We probably don't want to spend to much time improving it right now, because the `Coreg` classes will change and so will the code of this function.

In more details (**not much can be done right now, just to keep track for the future**), we would probably need to include:

- The `filtering` argument with the possibility to call classes of `filters.py` that could be combined with `Coreg` objects,
- Same for `biascorr.py` classes,
- Extract the statistics directly computed within the `Coreg` classes in `._metadata` (that needs to become a user-available attribute) to stay fully consistent with the results of the classes, instead of re-performing them manually here,
- Use the plots directly generated within the `Coreg` classes once those exist,
- Have an option for co-registration with point data once this exists,
- Use `Coreg.pipeline()` instead of performing the steps manually.

And, inversely, some parameters of this function should probably be moved to `Coreg` class so that they retain all flexibility:

- The `grid` argument should also be an option of `Coreg` objects, that always performs the same operation by default right now,
- Adding `exclude_mask`, `include_mask` might also be a great option, instead of just having an `inlier_mask` that requires projecting vectors on the same grid, but the mask passed should retain full flexibility (`Raster` object, `Vector` objects, or both).

In short, eventually, I think the one-liner co-registration needs to become a *convenience* function that passes the most robust default parameters to a `CoregPipeline` object and captures the outputs. All its options have to be available directly in the `Coreg` classes, while maintaining complete flexibility.
Do you picture it the same, @adehecq?

File: xdem/coreg.py [hunk 1]
```diff
@@ -1503,7 +1506,10 @@ def apply_matrix(
     :param invert: Invert the transformation matrix.
     :param centroid: The X/Y/Z transformation centroid. Irrelevant for pure translations. Defaults to the midpoint (Z=0)
     :param resampling: The resampling method to use. Can be `nearest`, `bilinear`, `cubic` or an integer from 0-5.
-    :param dilate_mask: Dilate the nan mask to exclude edge pixels that could be wrong.
+    :param dilate_mask: DEPRECATED - This option does not do anything anymore. Will be removed in the future.
```
[the remark this answers is not in the document]
Comment (adehecq) [reply xdem/coreg.py:1509]: I can't because it requires updating a lot of tests. This is why I raised an issue on this.

File: xdem/coreg.py [hunk 2]
```diff
@@ -2141,3 +2153,178 @@ def warp_dem(
...
+    :param out_dem_path: Path where to save the coregistered DEM. If set to None (default), will not save to file.
+    :param shpfile: path to a vector file containing areas to be masked for coregistration
+    :param coreg_method: The xdem coregistration method, or pipeline. If set to None, DEMs will be resampled to \
+ref grid and optionally filtered, but not coregistered. Will be used in priority over hmode and vmode.
+    :param hmode: The method to be used for horizontally aligning the DEMs, e.g. Nuth & Kaab or ICP. Can be any \
+of {list(vmodes_dict.keys())}.
+    :param vmode: The method to be used for vertically aligning the DEMs, e.g. mean/median bias correction or \
+deramping. Can be any of {list(hmodes_dict.keys())}.
+    :param deramp_degree: The degree of the polynomial for deramping.
+    :param grid: the grid to be used during coregistration, set either to "ref" or "src".
```
[the remark this answers is not in the document]
Comment (adehecq) [reply xdem/coreg.py:2204]: What would you use?

**Comment by adehecq** [comment]
> 
> In more details (**not much can be done right now, just to keep track for the future**), we would probably need to include:
> 
>     * The `filtering` argument with the possibility to call classes of `filters.py` that could be combined with `Coreg` objects,
> 
>     * Same for `biascorr.py` classes,

Good ideas yes.

> 
>     * Extract the statistics directly computed within the `Coreg` classes in `._metadata` (that needs to become a user-available attribute) to stay fully consistent with the results of the classes, instead of re-performing them manually here,
> 
>     * Use the plots directly generated within the `Coreg` classes once those exist,
> 
>     * Have an option for co-registration with point data once this exists,
> 

Yes, yes and yes.

>     * Use `Coreg.pipeline()` instead of performing the steps manually.
> 

There is an option to provide a Coreg.pipeline directly to `coreg_method`. The idea behind splitting the horizontal vs vertical alignment is that one might need to filter pixels differently (use a different inlier mask) for these two steps. I'm thinking for example about flat areas, that are not useful for Nuth & Kaab, but very useful for BiasCorr.

> 
> And, inversely, some parameters of this function should probably be moved to `Coreg` class so that they retain all flexibility:
> 
>     * The `grid` argument should also be an option of `Coreg` objects, that always performs the same operation by default right now,
> 
>     * Adding `exclude_mask`, `include_mask` might also be a great option, instead of just having an `inlier_mask` that requires projecting vectors on the same grid, but the mask passed should retain full flexibility (`Raster` object, `Vector` objects, or both).

Indeed, we could make the reprojection and creation of the `inlier_mask` part of the Coreg tools directly...

> 
> 
> In short, eventually, I think the one-liner co-registration needs to become a _convenience_ function that passes the most robust default parameters to a `CoregPipeline` object and captures the outputs. All its options have to be available directly in the `Coreg` classes, while maintaining complete flexibility. Do you picture it the same, @adehecq?

Yes, that seems sensible.

File: xdem/coreg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adehecq) [reply xdem/coreg.py:1509]: Solved with [60348f6](https://github.com/GlacioHack/xdem/pull/329/commits/60348f60873e7be836beffa8f870a83327672818)
